### ISSUE TITLE
PPOTrainer::train_step: support an auxiliary loss term

### ISSUE THREAD

**Issue by rjwalters** [rjwalters/thrust#2]
## Motivation

\`PPOTrainer::train_step\` currently composes its total loss as

\`\`\`
total = policy_loss + value_coef * value_loss - entropy_coef * entropy
\`\`\`

with no way for the caller to inject an additional term. Many algorithms need this:

- **Cross-agent redundancy penalty** (Slepian-Wolf MARL, P3 specialization): \`+ λ * Σ_{i<j} ||corr(Ẑ_i, Ẑ_j)||_F^2\`
- **Behavioural diversity** (DIAYN-style): \`- α * H(skill | trajectory)\`
- **Imitation regularizer**: \`+ β * KL(π_θ || π_demo)\`
- **Representation auxiliary**: \`+ γ * MSE(predicted_next_state, actual_next_state)\`

## Current API

\`\`\`rust
pub fn train_step<F>(
    &mut self,
    observations: &Tensor, actions: &Tensor,
    old_log_probs: &Tensor, old_values: &Tensor,
    advantages: &Tensor, returns: &Tensor,
    mut forward_fn: F,
) -> Result<TrainingStats>
where
    F: FnMut(&Tensor, &Tensor) -> (Tensor, Tensor, Tensor),
\`\`\`

## Proposal

Add an optional aux-loss callback:

\`\`\`rust
pub fn train_step_with_aux<F, A>(
    &mut self,
    // ...same args...
    mut forward_fn: F,
    mut aux_fn: A,
) -> Result<TrainingStats>
where
    F: FnMut(&Tensor, &Tensor) -> (Tensor, Tensor, Tensor),
    A: FnMut(&Tensor /* mb_obs */) -> Option<Tensor>, // scalar loss to add
\`\`\`

Or pass an extra \`aux_loss: Option<Tensor>\` along with \`(log_probs, entropy, values)\` from \`forward_fn\` (cleaner — keeps the loss computation co-located with the policy forward).

Either way, \`TrainingStats\` should grow an \`aux_loss\` field for logging.

## Notes

This is the natural counterpart to #1 (encoder tap): the tap exposes the features, this hook exposes the loss slot.

Backwards-compat path: keep \`train_step\` as-is, add \`train_step_with_aux\` and have the original delegate.

**Comment by rjwalters** [comment]
## Curator Note: Appears Already Implemented

Curating this issue per user request. Investigation finds that the proposal in this issue was implemented by commit `505e7f1` ("train/ppo: add auxiliary-loss hook to PPOTrainer"), which landed on `main` via the `slepian-wolf-marl-port` merge (`647a2f0`). The commit message explicitly references this issue ("Refs: #2").

### Acceptance criteria from the original proposal — verified against current `main`

| Proposal requirement | Status | Location |
|---|---|---|
| Add aux-loss-aware variant of `train_step` | Done | `src/train/ppo/trainer.rs:151` — `train_step_with_aux<F, A>` |
| Aux closure shape `FnMut(&Tensor /* mb_obs */) -> Option<Tensor>` | Done (exact match) | `src/train/ppo/trainer.rs:164` |
| Aux scalar added to total loss before backward | Done | `src/train/ppo/trainer.rs:244-246` |
| Backwards compat: `train_step` unchanged externally; delegates | Done | `src/train/ppo/trainer.rs:108-119` (no-op aux closure) |
| `TrainingStats` grows an `aux_loss` field for logging | Done | `src/train/ppo/stats.rs:27`; plumbed through `new`, `with_aux_loss`, `add`, `average` |
| Caller controls scaling (`λ * f(features)`) | Done by design | Trainer does not scale; comment at `trainer.rs:134-136` documents this |
| Tests | Done | `tests/test_ppo_aux_loss.rs` — three tests (no-aux yields zero `aux_loss`, aux is recorded, aux gradient measurably shrinks features) |

### Design choice note

The proposal listed two options: (a) a separate `aux_fn` callback, or (b) returning aux as a fourth tuple element from `forward_fn`. Option (a) was chosen and matches the API sketch in the proposal body verbatim. No follow-up work is required to switch approaches — option (a) is cleaner for the Slepian-Wolf cross-agent regularizer use case where aux loss is computed from features tapped via `MlpPolicy::encoder_features` (issue #1), since that tap can be invoked inside the aux closure independently of the policy forward.

### Companion issue

This is the loss-slot counterpart to #1 (encoder tap). Both shipped in the same merge: `encoder_features` landed in commit `049bbc4`. The pair is functional end-to-end — see `tests/test_ppo_aux_loss.rs:99-127` which exercises both together (aux closure calls `policy.encoder_features(mb_obs)` and applies a shrink penalty).

### Affected Files

- `src/train/ppo/trainer.rs` — new `train_step_with_aux`, delegation from `train_step`
- `src/train/ppo/stats.rs` — `aux_loss` field, `with_aux_loss` builder, averaging plumbing
- `tests/test_ppo_aux_loss.rs` — integration tests

### Test Plan

- [x] Manual verification: read `train_step_with_aux` implementation — aux scalar is added to `loss` before `loss.backward()` (`trainer.rs:241-252`)
- [x] Automated tests: three tests in `tests/test_ppo_aux_loss.rs` cover the no-aux delegation path, aux-recording path, and the gradient-actually-flows path
- [ ] Human verification: confirm the API ergonomics are acceptable as shipped (option-a callback vs. option-b tuple return)

### Recommended Action

This issue should be closed as already implemented. Per curator protocol I am not closing it; flagging with `loom:blocked` for human review and closure. If you want changes to the shipped API (e.g., switch to option-b tuple-return style, or expose aux scaling on `PPOConfig`), please open a follow-up issue with the specific request — that would be in-scope new work, not a continuation of #2.

**Comment by rjwalters** [comment]
Closing as completed — implemented in commit 505e7f1 (merged into main via 647a2f0).

`PPOTrainer::train_step_with_aux` exists at `src/train/ppo/trainer.rs`, takes an `aux_fn` closure on each minibatch's observations, and adds the optional scalar Tensor to total loss before `.backward()`. The original `train_step` delegates to it with a no-op aux closure (signature/behavior unchanged). `TrainingStats` got an `aux_loss` field plumbed through `.add`, `.average`, and `::with_aux_loss`.

Integration tests in `tests/test_ppo_aux_loss.rs` cover: (a) aux_loss == 0.0 when calling `train_step` directly, (b) aux_loss is recorded when calling `train_step_with_aux`, and (c) a large feature-shrink aux term measurably reduces feature mass.
